### ISSUE TITLE
Slow Convolutions on GTX 1080

### ISSUE THREAD

**Issue by ironbar** [Theano/Theano#4890]
Hi, 
I have been testing the new GTX 1080 and I have find that surprisingly convolution is much slower than GTX TITAN X. 
It is very strange because running the cuda nbody bechmark shows that GTX 1080 is faster than GTX TITAN X.
I have prepared a little benchmark with theano
https://github.com/ironbar/Theano_gpu_benchmark/blob/master/theanoGpuBenchmark.py

The first part computes the exp() of a vector, the second performs convolutions. In the first part 1080 is faster, in the second TITAN is faster. 
Here are the results.

```
Test                                                     1080    TITAN X
nBody Cuda Samples Bencharmk (GFLOPS)                    5133     3246
theano_benchmark_1 (exp(x)) (s)                           1.88   3.82
theano_benchmark_2 (convs)  (s)                           1.9     1.02
```

Could it be that theano is not ready for the new GTX 1080?
Maybe someone who has the 1080 could run the benchmark script and confirm my results.
In the same computer if I unplug the 1080 and plug the TITAN X everything works fine, so I guess the installation is ok.

Software configuration
Ubuntu 14.04
CUDA 8.0 RC
Driver 367.35
GTX 1080
Theano 0.9.0dev2.dev-74c0233b125798bab02300896f1e4c7a0900bb40

Thanks for your help
ironbar

**Comment by nouiz** [comment]
Did you installed cudnn? If not, I strongly recommend that you do it will
speed up CI involution by a lot. Make sure to use the latest cudnn version
5.1. it got speed up on the pascal GPU.

Also, check in Theano doc for time_once you can pass this to some Theano
flag to get faster conv for the 2nd and following call.

Fred

**Comment by ironbar** [comment]
Thanks a lot Fred, I was using cudnn 5, not 5.1. There's a big difference in speed!
Tomorrow I will post the benchmark  results with cudnn 5.1

**Comment by lamblin** [comment]
Also, can you check the content of the compiled function graph, to check that it is indeed using cudnn ops for convolution?
Also, like @nouiz said, maybe the default algorithm for convolution is slower, but `time_once` should select the best one.

**Comment by ironbar** [comment]
Hi, 
This is the speedup on GTX 1080 using cudnn 5.1 instead of cudnn 5


```
Test                                   1080_cudnn5    TITAN X      1080_cudnn5.1     1080_cudnn5.1_opt
nBody Cuda Samples Bencharmk (GFLOPS)      5133        3246            5133             5133
theano_benchmark_1 (exp(x)) (s)            1.88        3.82            1.9               1.9
theano_benchmark_2 (convs)  (s)            1.9         1.02            1.24               0.97
```

I have tested `time_once` but the effect on my computer is very small. 
In the last test I used the following flag `THEANO_FLAGS=optimizer_including=cudnn`, and that had a considerable effect on the benchmark time.

Thanks for your help
ironbar

**Comment by nouiz** [comment]
Which version of Theano do you use? The dev version?

In this line

theano_benchmark_2 (convs)  (s)            1.9         1.02
1.24               0.97


We see that the opt did a significative speed up and that cudnn 5.1 did a
good speed up too.

The last number 0.97, which flag you add? time_once? Or just the one you
wrote? Can you give exactly the time_once flags you used to be sure you got
the right one?

Also, the first call will be much slower with time_once then with others.
This is normal, you should remove it from the timing.

**Comment by ironbar** [comment]
Hi, 
I'm using the theano dev version "Theano 0.9.0dev2.dev-74c0233b125798bab02300896f1e4c7a0900bb40"

In 0.97 I used both `time_once` and `THEANO_FLAGS=optimizer_including=cudnn`, but removing 'time_once' had no effect. 

I have made the first call out of the loop but again in this example the effect is negligible. Maybe in other cases it helps. 

I set the configuration of `time_once` in the .theanorc file 
```
[dnn.conv]
algo_fwd = time_once
algo_bwd_data = time_once
```
